numba/tests/test_record_dtype.py

```diff
@@ -1371,23 +1368,260 @@ def foo(rec, flag=None):
+        for arg in [nbarr, nbarr[0]]:
+            ty = typeof(arg)
+            pyfunc = recarray_getitem_field_return2_2d
+            arr_expected = pyfunc(arg)
+            cfunc = self.get_cfunc(pyfunc, (ty,))
+            arr_res = cfunc(arg)
+            np.testing.assert_equal(arr_res, arr_expected)
+
+    def test_return_getattr_getitem_fieldname(self):
+        # Test __getitem__ with field name and getattr .field_name
+        #
+        # This tests returning a array of nestedarrays when passing an array and
+        # returning a nestedarray when passing a record
+
+        nbarr = np.recarray(2, dtype=recordwitharray)
+        nbarr[0] = np.array([(1, (2,3))], dtype=recordwitharray)[0]
+        for arg in [nbarr, nbarr[0]]:
+            for pyfunc in [recarray_getitem_field_return,
+                           recarray_getitem_field_return2]:
+                ty = typeof(arg)
+                arr_expected = pyfunc(arg)
+                cfunc = self.get_cfunc(pyfunc, (ty,))
+                arr_res = cfunc(arg)
+                np.testing.assert_equal(arr_res, arr_expected)
+
+    def test_return_array(self):
+        # Test getitem record AND array within record and returning it
+
+        nbval = np.recarray(2, dtype=recordwitharray)
+        nbval[0] = np.array([(1, (2,3))], dtype=recordwitharray)[0]
+        for pyfunc in [record_read_array0, record_read_array0_alt]:
+            arr_expected = pyfunc(nbval)
+            cfunc = self.get_cfunc(pyfunc, (typeof(nbval),))
+            arr_res = cfunc(nbval)
+            np.testing.assert_equal(arr_expected, arr_res)
+
+    def test_slice_2d_array(self):
+        # test slicing the nestedarray inside a record
+        nbarr = np.recarray(2, dtype=recordwith2darray)
+        nbarr[0] = np.array([(1, ((1,2),(4,5),(2,3)))],
+                            dtype=recordwith2darray)[0]
+
+        funcs = rec_getitem_field_slice_2d, recarray_getitem_field_slice_2d
+        for arg, pyfunc in zip([nbarr[0], nbarr], funcs):
+            ty = typeof(arg)
+            arr_expected = pyfunc(arg)
+            cfunc = self.get_cfunc(pyfunc, (ty,))
+            arr_res = cfunc(arg)
+            np.testing.assert_equal(arr_res, arr_expected)
+
+    def test_shape(self):
+        # test getting the shape of a nestedarray inside a record
+        nbarr = np.recarray(2, dtype=recordwith2darray)
+        nbarr[0] = np.array([(1, ((1,2),(4,5),(2,3)))],
+                            dtype=recordwith2darray)[0]
+
+        arg = nbarr[0]
+        pyfunc = get_shape
+        ty = typeof(arg)
+        arr_expected = pyfunc(arg)
+        cfunc = self.get_cfunc(pyfunc, (ty,))
+        arr_res = cfunc(arg)
+        np.testing.assert_equal(arr_res, arr_expected)
+
+    def test_corner_slice(self):
+        # testing corner cases while slicing nested arrays
+        nbarr = np.recarray((1, 2, 3, 5, 7, 13, 17), dtype=recordwith4darray,
+                            order='F')
```

Comment:
Thanks for adding these somewhat unusual test cases as per our discussion at the public meeting, much appreciated.